<a href="https://colab.research.google.com/github/anoohyabhaskarla/INFO_5731/blob/main/Bhaskarka_Anoohya_Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [19]:
# Your code here
import math

def calculate_hemisphere_volume(radius):
    volume = (2/3) * math.pi * radius**3
    return volume
print("r    V")

for r in range(2, 21, 2):
    volume = calculate_hemisphere_volume(r)
    print(f"{r}    {volume}")



r    V
2    16.755160819145562
4    134.0412865531645
6    452.38934211693015
8    1072.330292425316
10    2094.3951023931954
12    3619.114736935441
14    5747.0201609669275
16    8578.642339402528
18    12214.512237157114
20    16755.160819145563


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [20]:
# Write code for each of the sub parts with proper comments.
import random

def generate_random_sequence(length, minimum, maximum):
    return [random.randint(minimum, maximum) for _ in range(length)]

def calculate_average(numbers):
    return sum(numbers) / len(numbers)

# Set the parameters
sequence_length = 12
minimum_value = 50
maximum_value = 150

# (a) Generate the random sequence
random_sequence = generate_random_sequence(sequence_length, minimum_value, maximum_value)

# (b) Sort the sequence in ascending order
ascending_sorted_sequence = sorted(random_sequence)
print("Sorted Random Sequence (Ascending):", ascending_sorted_sequence)

# (c) Sort the sequence in descending order
descending_sorted_sequence = sorted(random_sequence, reverse=True)
print("Sorted Random Sequence (Descending):", descending_sorted_sequence)

# (d) Compute and output the average of the entire group of generated numbers
average_whole_group = calculate_average(random_sequence)
print("Average of the entire group:", average_whole_group)

# (e) Compute and display the average of the first 6 numbers from both ascending and descending sorted lists
average_first_six_ascending = calculate_average(ascending_sorted_sequence[:6])
average_first_six_descending = calculate_average(descending_sorted_sequence[:6])

print("Average of the first 6 numbers (Ascending):", average_first_six_ascending)
print("Average of the first 6 numbers (Descending):", average_first_six_descending)



Sorted Random Sequence (Ascending): [58, 76, 97, 99, 103, 104, 107, 109, 113, 123, 142, 146]
Sorted Random Sequence (Descending): [146, 142, 123, 113, 109, 107, 104, 103, 99, 97, 76, 58]
Average of the entire group: 106.41666666666667
Average of the first 6 numbers (Ascending): 89.5
Average of the first 6 numbers (Descending): 123.33333333333333


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [21]:
# Your code here
import string

def get_user_input():
    sentence1 = input("Enter the first sentence: ")
    sentence2 = input("Enter the second sentence: ")
    return sentence1, sentence2

def transform_to_word_list(sentence):
    words = sentence.split()
    # Remove punctuation from each word
    words = [word.strip(string.punctuation) for word in words]
    return words

def find_common_words(list1, list2):
    return list(set(list1) & set(list2))

def remove_punctuation(sentence):
    return sentence.translate(str.maketrans("", "", string.punctuation))

def count_vowels(sentence):
    vowels = "aeiouAEIOU"
    vowel_count = {vowel: sentence.lower().count(vowel) for vowel in vowels}
    return vowel_count

# (a) Prompt the user to input two sentences.
sentence1, sentence2 = get_user_input()

# (b) Transform each sentence into a list of words.
word_list1 = transform_to_word_list(sentence1)
word_list2 = transform_to_word_list(sentence2)

# Print the transformed word lists
print("Transformed Word List 1:", word_list1)
print("Transformed Word List 2:", word_list2)

# (c) Compare the two sentences. Print out a list of words that occur in both sentences.
common_words = find_common_words(word_list1, word_list2)
print("Common Words:", common_words)

# (d) Remove any punctuation from the sentences.
sanitized_sentence1 = remove_punctuation(sentence1)
sanitized_sentence2 = remove_punctuation(sentence2)
print("Sanitized Sentence 1:", sanitized_sentence1)
print("Sanitized Sentence 2:", sanitized_sentence2)

# (e) Count the number of vowels in each sentence.
vowel_count_sentence1 = count_vowels(sentence1)
vowel_count_sentence2 = count_vowels(sentence2)
print("Vowel Count Sentence 1:", vowel_count_sentence1)
print("Vowel Count Sentence 2:", vowel_count_sentence2)



Enter the first sentence: who are you
Enter the second sentence: I am anoohya
Transformed Word List 1: ['who', 'are', 'you']
Transformed Word List 2: ['I', 'am', 'anoohya']
Common Words: []
Sanitized Sentence 1: who are you
Sanitized Sentence 2: I am anoohya
Vowel Count Sentence 1: {'a': 1, 'e': 1, 'i': 0, 'o': 2, 'u': 1, 'A': 0, 'E': 0, 'I': 0, 'O': 0, 'U': 0}
Vowel Count Sentence 2: {'a': 3, 'e': 0, 'i': 1, 'o': 2, 'u': 0, 'A': 0, 'E': 0, 'I': 0, 'O': 0, 'U': 0}


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
# Your code here
# (a) Write program that takes a mathematical expression as input and evaluates it.
#The program should support addition, subtraction, multiplication, and division.

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Simple Calculator!")
    print("Supported operations: +, -, *, /")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break

        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()


# (b) Extend the calculator to handle parentheses in mathematical expressions.
# Implement the necessary logic to evaluate expressions with nested parentheses.

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, parentheses")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

#(c) Enhance your calculator to include additional mathematical operations such as exponentiation,
#square root, and modulus.

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()


# (d) Implement error handling in the calculator program.
# Handle cases such as division by zero and invalid input expressions.
#Provide meaningful error messages

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except ValueError:
        return "Error: Invalid input expression"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break
        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

#(e) Implement a command to exit the calculator.

import math

def evaluate_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero"
    except ValueError:
        return "Error: Invalid input expression"
    except Exception as e:
        return f"Error: {e}"

def main():
    print("Welcome to the Enhanced Calculator!")
    print("Supported operations: +, -, *, /, ^ (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        user_input = input("Enter a mathematical expression: ")

        if user_input.lower() == "exit":
            print("Exiting the calculator. Goodbye!")
            break

        result = evaluate_expression(user_input)
        print(f"Result: {result}")

if __name__ == "__main__":
    main()

Welcome to the Simple Calculator!
Supported operations: +, -, *, /
Type 'exit' to quit.
Result: 11
Result: 2
Result: 63
Result: 12.5
Result: 4.0
Exiting the calculator. Goodbye!
Welcome to the Enhanced Calculator!
Supported operations: +, -, *, /, parentheses
Type 'exit' to quit.


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [1]:
import pandas as pd
import statistics

# a) Function to compute year-over-year differences in occupied housing units
def compute_yearly_differences(data):
    occupied_data = data[data['title_field'] == 'Occupied Housing Units']
    occupied_data = occupied_data[['year', 'value']].sort_values(by='year')
    differences = occupied_data['value'].diff().dropna()
    sorted_differences = differences.sort_values()

    print("\nSorted Yearly Differences in Occupied Housing Units:")
    for i, diff in enumerate(sorted_differences, start=1):
        print(f"{occupied_data.iloc[i-1]['year']}-{occupied_data.iloc[i]['year']}: {diff}")
    return sorted_differences

# b) Function to calculate and display the percentage of vacant vs. occupied housing units
def calculate_unit_percentages(data):
    years = data['year'].unique()
    print(f"\n{'Year':<10} {'Vacant Units (%)':<20} {'Occupied Units (%)':<25}")
    for year in years:
        total_units = data[data['year'] == year]['value'].sum()
        vacant_units = data[(data['year'] == year) & (data['title_field'] == 'Vacant Housing Units')]['value'].values[0]
        occupied_units = data[(data['year'] == year) & (data['title_field'] == 'Occupied Housing Units')]['value'].values[0]
        vacant_pct = (vacant_units / total_units) * 100
        occupied_pct = (occupied_units / total_units) * 100
        print(f"{year:<10} {vacant_pct:<20.2f} {occupied_pct:<25.2f}")

# c) Function to find and print the year with maximum vacant and occupied units
def max_vacant_occupied_year(data):
    vacant_data = data[data['title_field'] == 'Vacant Housing Units']
    occupied_data = data[data['title_field'] == 'Occupied Housing Units']
    max_vacant_year = vacant_data.loc[vacant_data['value'].idxmax()]
    max_occupied_year = occupied_data.loc[occupied_data['value'].idxmax()]

    print("\nYear with Maximum Vacant and Occupied Units:")
    print(f"{'Category':<25} {'Year':<10} {'Units':<10}")
    print(f"Vacant Units               {max_vacant_year['year']:<10} {max_vacant_year['value']:<10}")
    print(f"Occupied Units             {max_occupied_year['year']:<10} {max_occupied_year['value']:<10}")

# d) Function to compute the harmonic mean of total housing units
def harmonic_mean_total_units(data):
    yearly_totals = data.groupby('year')['value'].sum()
    harmonic_mean = statistics.harmonic_mean(yearly_totals)
    print(f"\nHarmonic Mean of Total Housing Units: {harmonic_mean:.2f}")

# Main function to load data and call the relevant analysis functions
def main():
    # Loading data from the provided CSV file
    data = pd.read_csv('Assignment1_denton_housing.csv')

    # (a) Compute and print year-over-year differences in occupied housing units
    compute_yearly_differences(data)

    # (b) Calculate and display the percentage of vacant and occupied housing units by year
    calculate_unit_percentages(data)

    # (c) Find the year with the highest number of vacant and occupied housing units
    max_vacant_occupied_year(data)

    # (d) Compute the harmonic mean of total housing units over the years
    harmonic_mean_total_units(data)

if __name__ == "__main__":
    main()




Sorted Yearly Differences in Occupied Housing Units:
2008-2009: -795.0
2009-2010: -303.0
2010-2011: 1404.0
2011-2012: 1565.0
2012-2013: 2622.0
2013-2014: 5091.0

Year       Vacant Units (%)     Occupied Units (%)       
2014       1.94                 31.84                    
2013       3.00                 31.12                    
2012       2.26                 31.61                    
2011       3.21                 30.99                    
2010       2.77                 31.29                    
2009       3.10                 31.05                    
2008       3.74                 30.63                    

Year with Maximum Vacant and Occupied Units:
Category                  Year       Units     
Vacant Units               2008       4483      
Occupied Units             2014       46295     

Harmonic Mean of Total Housing Units: 130257.98


# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

# **I just want to learn how this program works in general. So I just splitup the question accordingly.**

Prompt 1: Create a Python program for a product management organization that keeps track of inventory. It should be possible to add, remove, and update products via the system. Keep track of product information, including Supplier, Price, Quantity, Category, Name, and Product ID. To preserve data, use a CSV file. Put methods in place to import and save inventory data from CSV files.

In [11]:
import csv
import os
from collections import defaultdict

# File to store inventory
data_file = "inventory.csv"

# Default stock threshold for restocking
RESTOCK_THRESHOLD = 10
RESTOCK_QUANTITY = 50

# Sample initial data (50 products across categories)
def initialize_inventory():
    if not os.path.exists(data_file) or os.stat(data_file).st_size == 0:
        with open(data_file, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Quantity", "Price"])
            sample_data = [
                [1, "Laptop", "Electronics", 15, 1000],
                [2, "Mouse", "Electronics", 5, 25],
                [3, "Keyboard", "Electronics", 7, 45],
                [4, "Monitor", "Electronics", 12, 150],
                [5, "Headphones", "Electronics", 9, 75],
                [6, "Smartphone", "Electronics", 20, 800],
                [7, "Tablet", "Electronics", 3, 600],
                [8, "TV", "Electronics", 5, 900],
                [9, "Speaker", "Electronics", 18, 120],
                [10, "Charger", "Electronics", 25, 30],
                [11, "Sofa", "Furniture", 5, 700],
                [12, "Chair", "Furniture", 8, 150],
                [13, "Table", "Furniture", 10, 300],
                [14, "Shelf", "Furniture", 4, 400],
                [15, "Bed", "Furniture", 2, 1000],
                [16, "Lamp", "Furniture", 15, 50],
                [17, "Pillow", "Furniture", 30, 25],
                [18, "Blanket", "Furniture", 12, 80],
                [19, "Wardrobe", "Furniture", 6, 900],
                [20, "Mattress", "Furniture", 7, 500],
                [21, "Shampoo", "Grocery", 25, 10],
                [22, "Soap", "Grocery", 20, 5],
                [23, "Toothpaste", "Grocery", 30, 8],
                [24, "Detergent", "Grocery", 12, 20],
                [25, "Milk", "Grocery", 50, 3],
                [26, "Eggs", "Grocery", 40, 2],
                [27, "Rice", "Grocery", 15, 25],
                [28, "Sugar", "Grocery", 18, 15],
                [29, "Oil", "Grocery", 10, 50],
                [30, "Flour", "Grocery", 22, 10],
                [31, "Jeans", "Clothing", 14, 60],
                [32, "T-Shirt", "Clothing", 25, 30],
                [33, "Jacket", "Clothing", 10, 100],
                [34, "Sweater", "Clothing", 8, 80],
                [35, "Shoes", "Clothing", 12, 120],
                [36, "Cap", "Clothing", 20, 15],
                [37, "Socks", "Clothing", 30, 10],
                [38, "Gloves", "Clothing", 7, 40],
                [39, "Scarf", "Clothing", 18, 25],
                [40, "Watch", "Clothing", 5, 200],
            ]
            writer.writerows(sample_data)

def read_inventory():
    try:
        with open(data_file, mode='r') as file:
            reader = csv.reader(file)
            next(reader)  # Skip header
            return [row for row in reader]
    except Exception as e:
        print("Error reading inventory file:", e)
        return []

def write_inventory(data):
    with open(data_file, mode='w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(["ID", "Name", "Category", "Quantity", "Price"])
        writer.writerows(data)

def add_product(name, category, quantity, price):
    data = read_inventory()
    new_id = len(data) + 1
    data.append([new_id, name, category, quantity, price])
    write_inventory(data)
    print("Product added successfully!")

def delete_product(product_id):
    data = read_inventory()
    data = [row for row in data if int(row[0]) != product_id]
    write_inventory(data)
    print("Product deleted successfully!")

def update_product(product_id, name=None, category=None, quantity=None, price=None):
    data = read_inventory()
    for row in data:
        if int(row[0]) == product_id:
            if name:
                row[1] = name
            if category:
                row[2] = category
            if quantity is not None:
                row[3] = quantity
            if price is not None:
                row[4] = price
    write_inventory(data)
    print("Product updated successfully!")

def search_product(keyword):
    data = read_inventory()
    return [row for row in data if keyword.lower() in row[1].lower()]

def generate_report():
    data = read_inventory()
    category_totals = defaultdict(float)
    for row in data:
        category_totals[row[2]] += float(row[4]) * int(row[3])
    for category, total in category_totals.items():
        print(f"Category: {category} - Total Value: ${total:.2f}")

def automate_restock():
    data = read_inventory()
    for row in data:
        if int(row[3]) < RESTOCK_THRESHOLD:
            row[3] = RESTOCK_QUANTITY  # Restock to full
    write_inventory(data)
    print("Restocking completed!")

# Initialize inventory file
initialize_inventory()


Prompt2:  why theer's no display of the output though it runs?

In [13]:
def initialize_inventory():
    if not os.path.exists(data_file) or os.stat(data_file).st_size == 0:
        with open(data_file, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Quantity", "Price"])
            sample_data = ['inventory.csv' ]
            writer.writerows(sample_data)
        print("Inventory initialized successfully!")
print(read_inventory())
generate_report()


[['001', 'Laptop', 'Electronics', '10', '1200', 'TechCorp', 'NO'], ['002', 'Smartphone', 'Electronics', '25', '800', 'MobileMakers', 'NO']]
Category: Electronics - Total Value: $32000.00


Prompt 3: Improve the inventory system by incorporating search capabilities. Give customers the option to look for products by Product ID, Name, or Category. To increase usability, use partial matches and case-insensitive searches. Make sure the system responds meaningfully in the event that a product cannot be located.

In [7]:
import csv

class InventoryManager:
    def __init__(self, filename="inventory.csv"):
        self.filename = filename
        self.inventory = []
        self.load_inventory()

    def load_inventory(self):
        """Load inventory data from a CSV file."""
        try:
            with open(self.filename, mode='r', newline='') as file:
                reader = csv.DictReader(file)
                self.inventory = list(reader)
        except FileNotFoundError:
            self.inventory = []
        except Exception as e:
            print(f"Error loading inventory: {e}")

    def save_inventory(self):
        """Save inventory data to a CSV file."""
        with open(self.filename, mode='w', newline='') as file:
            fieldnames = ["Product ID", "Name", "Category", "Quantity", "Price", "Supplier"]
            writer = csv.DictWriter(file, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(self.inventory)

    def add_product(self, product_id, name, category, quantity, price, supplier):
        """Add a new product to the inventory."""
        self.inventory.append({
            "Product ID": product_id,
            "Name": name,
            "Category": category,
            "Quantity": quantity,
            "Price": price,
            "Supplier": supplier
        })
        self.save_inventory()

    def delete_product(self, product_id):
        """Delete a product by Product ID."""
        self.inventory = [p for p in self.inventory if p["Product ID"] != product_id]
        self.save_inventory()

    def update_product(self, product_id, field, new_value):
        """Update a product's field with a new value."""
        for product in self.inventory:
            if product["Product ID"] == product_id:
                product[field] = new_value
                break
        self.save_inventory()

    def search_product(self, search_term):
        """Search for products by Product ID, Name, or Category (case-insensitive, partial match)."""
        search_term = search_term.lower()
        results = [p for p in self.inventory if search_term in p["Product ID"].lower() or
                   search_term in p["Name"].lower() or
                   search_term in p["Category"].lower()]

        if results:
            for product in results:
                print(product)
        else:
            print("No products found matching the search criteria.")

# Example Usage
if __name__ == "__main__":
    inventory = InventoryManager()
    inventory.add_product("001", "Laptop", "Electronics", "10", "1200", "TechCorp")
    inventory.add_product("002", "Smartphone", "Electronics", "25", "800", "MobileMakers")
    inventory.search_product("laptop")  # Should return the laptop entry
    inventory.search_product("Electronics")  # Should return all electronics


{'Product ID': '001', 'Name': 'Laptop', 'Category': 'Electronics', 'Quantity': '10', 'Price': '1200', 'Supplier': 'TechCorp'}
{'Product ID': '001', 'Name': 'Laptop', 'Category': 'Electronics', 'Quantity': '10', 'Price': '1200', 'Supplier': 'TechCorp'}
{'Product ID': '002', 'Name': 'Smartphone', 'Category': 'Electronics', 'Quantity': '25', 'Price': '800', 'Supplier': 'MobileMakers'}


Prompt 4:Now, Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

In [2]:
import csv
import os

# Constants
CSV_FILE = "inventory.csv"
LOW_STOCK_THRESHOLD = 5  # Auto-restock threshold

class InventoryManager:
    def __init__(self, filename=CSV_FILE):
        self.filename = filename
        self.fields = ["ID", "Name", "Category", "Quantity", "Price"]
        self.load_inventory()

    def load_inventory(self):
        """Loads or initializes inventory from a CSV file, handling corruption."""
        if not os.path.exists(self.filename) or os.stat(self.filename).st_size == 0:
            print("Initializing new inventory...")
            self.initialize_inventory()
        else:
            try:
                with open(self.filename, mode="r", newline="", encoding="utf-8") as file:
                    reader = csv.DictReader(file)
                    self.inventory = [row for row in reader]
                    if not self.inventory or any(field not in self.inventory[0] for field in self.fields):
                        raise ValueError("Corrupted CSV structure.")
            except Exception as e:
                print(f"Error loading inventory: {e}. Reinitializing...")
                self.initialize_inventory()

    def initialize_inventory(self):
        """Initializes the inventory with 50 preloaded products."""
        self.inventory = [
            {"ID": str(i+1), "Name": f"Product{i+1}", "Category": ["Electronics", "Clothing", "Grocery", "Furniture", "Toys"][i % 5],
             "Quantity": str(10 + i % 5), "Price": str(round(5 + (i % 5) * 10, 2))}
            for i in range(50)
        ]
        self.save_inventory()

    def save_inventory(self):
        """Saves the inventory to the CSV file."""
        with open(self.filename, mode="w", newline="", encoding="utf-8") as file:
            writer = csv.DictWriter(file, fieldnames=self.fields)
            writer.writeheader()
            writer.writerows(self.inventory)

    def display_inventory(self):
        """Displays all inventory items."""
        print("\nCurrent Inventory:")
        for item in self.inventory:
            print(f"{item['ID']}: {item['Name']} ({item['Category']}) - {item['Quantity']} units @ ${item['Price']}")

    def add_product(self, name, category, quantity, price):
        """Adds a new product."""
        new_id = str(len(self.inventory) + 1)
        self.inventory.append({"ID": new_id, "Name": name, "Category": category, "Quantity": str(quantity), "Price": str(price)})
        self.save_inventory()
        print(f"Added {name} with ID {new_id}.")

    def delete_product(self, product_id):
        """Deletes a product by ID."""
        self.inventory = [item for item in self.inventory if item["ID"] != product_id]
        self.save_inventory()
        print(f"Deleted product ID {product_id}.")

    def update_product(self, product_id, field, new_value):
        """Updates product details."""
        for item in self.inventory:
            if item["ID"] == product_id:
                item[field] = new_value
                self.save_inventory()
                print(f"Updated product ID {product_id} {field} to {new_value}.")
                return
        print(f"Product ID {product_id} not found.")

    def search_product(self, query, field="Name"):
        """Searches for a product by name, category, or ID."""
        results = [item for item in self.inventory if query.lower() in item[field].lower()]
        if results:
            print("\nSearch Results:")
            for item in results:
                print(f"{item['ID']}: {item['Name']} ({item['Category']}) - {item['Quantity']} units @ ${item['Price']}")
        else:
            print("No matching products found.")

    def auto_restock(self):
        """Automatically restocks low-stock items."""
        for item in self.inventory:
            if int(item["Quantity"]) < LOW_STOCK_THRESHOLD:
                item["Quantity"] = str(LOW_STOCK_THRESHOLD + 5)  # Restock to threshold + buffer
                print(f"Auto-restocked {item['Name']} (ID {item['ID']}) to {item['Quantity']} units.")
        self.save_inventory()

    def generate_report(self):
        """Generates an inventory value and category breakdown report."""
        total_value = 0
        category_counts = {}

        for item in self.inventory:
            total_value += int(item["Quantity"]) * float(item["Price"])
            category_counts[item["Category"]] = category_counts.get(item["Category"], 0) + int(item["Quantity"])

        print("\nInventory Report:")
        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, count in category_counts.items():
            print(f"  - {category}: {count} units")

# Main execution
if __name__ == "__main__":
    manager = InventoryManager()

    while True:
        print("\nInventory Management System")
        print("1. Display Inventory")
        print("2. Add Product")
        print("3. Delete Product")
        print("4. Update Product")
        print("5. Search Product")
        print("6. Auto-Restock Low Stock Items")
        print("7. Generate Report")
        print("8. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            manager.display_inventory()
        elif choice == "2":
            name = input("Enter product name: ").strip()
            category = input("Enter product category: ").strip()
            quantity = input("Enter quantity: ").strip()
            price = input("Enter price: ").strip()
            manager.add_product(name, category, quantity, price)
        elif choice == "3":
            product_id = input("Enter product ID to delete: ").strip()
            manager.delete_product(product_id)
        elif choice == "4":
            product_id = input("Enter product ID to update: ").strip()
            field = input("Enter field to update (Name/Category/Quantity/Price): ").strip()
            new_value = input(f"Enter new value for {field}: ").strip()
            manager.update_product(product_id, field, new_value)
        elif choice == "5":
            query = input("Enter product name/category/ID to search: ").strip()
            field = input("Search by (Name/Category/ID): ").strip()
            manager.search_product(query, field)
        elif choice == "6":
            manager.auto_restock()
        elif choice == "7":
            manager.generate_report()
        elif choice == "8":
            print("Exiting system...")
            break
        else:
            print("Invalid choice. Please enter a number from 1 to 8.")



Inventory Management System
1. Display Inventory
2. Add Product
3. Delete Product
4. Update Product
5. Search Product
6. Auto-Restock Low Stock Items
7. Generate Report
8. Exit
Enter your choice: 7

Inventory Report:
Total Inventory Value: $1815.00
Category Breakdown:
  - Electronics: 20 units
  - Clothing: 22 units
  - Grocery: 12 units
  - Furniture: 13 units
  - Toys: 14 units

Inventory Management System
1. Display Inventory
2. Add Product
3. Delete Product
4. Update Product
5. Search Product
6. Auto-Restock Low Stock Items
7. Generate Report
8. Exit
Enter your choice: 8
Exiting system...


Prompt 5: Let me know how to identify the edge cases from the above code.

To identify edge cases in your inventory management system, analyze input constraints like numeric-only Product IDs and non-empty product names. Check for duplicates by preventing repeated Product IDs or names. Handle invalid data types by ensuring stock and price fields don’t contain text or special characters. Test boundary values such as negative stock or IDs starting with zero. Verify search functionality with partial and case-sensitive queries and confirm reports generate product-level details. Attempt operations on non-existent products to ensure proper error handling. Simulate corrupted CSV files to test system resilience. Validate auto-restocking to prevent unintended stock levels, ensure deletion prompts prevent accidental removals, and test unexpected shutdowns to maintain data integrity.

Prompt 6: Now, Taking these soem of the important cases in consideration like Duplicate entries should be prevented in the CSV file, and a product must not be assigned to multiple categories. Product IDs should strictly contain only numbers and must not begin with zero. Additionally, product names cannot be left empty, and stock quantities must always be whole numbers, disallowing negative or decimal values. If a user attempts to update or delete a non-existent product, they should receive a notification. Moreover, report generation should support an additional level to allow product-specific breakdowns when needed. Lastly, the system should block adding products with an already existing Product ID.How to improve the code?

In [5]:
import csv
import os

# Constants
CSV_FILE = "inventory.csv"
LOW_STOCK_THRESHOLD = 5  # Auto-restock threshold

class InventoryManager:
    def __init__(self, filename=CSV_FILE):
        self.filename = filename
        self.fields = ["ID", "Name", "Category", "Quantity", "Price"]
        self.load_inventory()

    def load_inventory(self):
        """Loads or initializes inventory from a CSV file, handling corruption."""
        if not os.path.exists(self.filename) or os.stat(self.filename).st_size == 0:
            print("Initializing new inventory...")
            self.initialize_inventory()
        else:
            try:
                with open(self.filename, mode="r", newline="", encoding="utf-8") as file:
                    reader = csv.DictReader(file)
                    self.inventory = [row for row in reader]
                    if not self.inventory or any(field not in self.inventory[0] for field in self.fields):
                        raise ValueError("Corrupted CSV structure.")
            except Exception as e:
                print(f"Error loading inventory: {e}. Reinitializing...")
                self.initialize_inventory()

    def initialize_inventory(self):
        """Initializes the inventory with 50 preloaded products."""
        self.inventory = []
        for i in range(50):
            product_id = str(i + 1)
            name = f"Product{i+1}"
            category = ["Electronics", "Clothing", "Grocery", "Furniture", "Toys"][i % 5]
            quantity = str(10 + i % 5)
            price = str(round(5 + (i % 5) * 10, 2))
            self.inventory.append({"ID": product_id, "Name": name, "Category": category, "Quantity": quantity, "Price": price})
        self.save_inventory()

    def save_inventory(self):
        """Saves the inventory to the CSV file."""
        with open(self.filename, mode="w", newline="", encoding="utf-8") as file:
            writer = csv.DictWriter(file, fieldnames=self.fields)
            writer.writeheader()
            writer.writerows(self.inventory)

    def is_valid_product_id(self, product_id):
        """Ensures Product ID is numeric and does not start with zero."""
        return product_id.isdigit() and not product_id.startswith("0")

    def is_valid_quantity(self, quantity):
        """Ensures stock quantity is a whole number and non-negative."""
        return quantity.isdigit() and int(quantity) >= 0

    def add_product(self, name, category, quantity, price):
        """Adds a new product, ensuring uniqueness and valid values."""
        if not name.strip():
            print("Product name cannot be empty.")
            return
        if not self.is_valid_quantity(quantity):
            print("Quantity must be a whole number and non-negative.")
            return
        if any(item["Name"].lower() == name.lower() for item in self.inventory):
            print("Product with this name already exists.")
            return
        new_id = str(len(self.inventory) + 1)
        self.inventory.append({"ID": new_id, "Name": name, "Category": category, "Quantity": quantity, "Price": price})
        self.save_inventory()
        print(f"Added {name} with ID {new_id}.")

    def delete_product(self, product_id):
        """Deletes a product by ID if it exists."""
        if not any(item["ID"] == product_id for item in self.inventory):
            print("Product ID not found.")
            return
        self.inventory = [item for item in self.inventory if item["ID"] != product_id]
        self.save_inventory()
        print(f"Deleted product ID {product_id}.")

    def update_product(self, product_id, field, new_value):
        """Updates product details with validation."""
        for item in self.inventory:
            if item["ID"] == product_id:
                if field == "Quantity" and not self.is_valid_quantity(new_value):
                    print("Quantity must be a whole number and non-negative.")
                    return
                item[field] = new_value
                self.save_inventory()
                print(f"Updated product ID {product_id} {field} to {new_value}.")
                return
        print("Product ID not found.")

    def generate_report(self, specific_product_id=None):
        """Generates an inventory value and category breakdown report. Optionally includes product-specific breakdown."""
        total_value = 0
        category_counts = {}

        if specific_product_id:
            product = next((item for item in self.inventory if item["ID"] == specific_product_id), None)
            if product:
                print(f"Product Report for {product['Name']} (ID {product['ID']}):")
                print(f"  Category: {product['Category']}")
                print(f"  Quantity: {product['Quantity']} units")
                print(f"  Price: ${product['Price']}")
                return
            else:
                print("Product ID not found.")
                return

        for item in self.inventory:
            total_value += int(item["Quantity"]) * float(item["Price"])
            category_counts[item["Category"]] = category_counts.get(item["Category"], 0) + int(item["Quantity"])

        print("\nInventory Report:")
        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, count in category_counts.items():
            print(f"  - {category}: {count} units")

# Main execution
if __name__ == "__main__":
    manager = InventoryManager()

    while True:
        print("\nInventory Management System")
        print("1. Display Inventory")
        print("2. Add Product")
        print("3. Delete Product")
        print("4. Update Product")
        print("5. Generate Report")
        print("6. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            manager.display_inventory()
        elif choice == "2":
            name = input("Enter product name: ").strip()
            category = input("Enter product category: ").strip()
            quantity = input("Enter quantity: ").strip()
            price = input("Enter price: ").strip()
            manager.add_product(name, category, quantity, price)
        elif choice == "3":
            product_id = input("Enter product ID to delete: ").strip()
            manager.delete_product(product_id)
        elif choice == "4":
            product_id = input("Enter product ID to update: ").strip()
            field = input("Enter field to update (Name/Category/Quantity/Price): ").strip()
            new_value = input(f"Enter new value for {field}: ").strip()
            manager.update_product(product_id, field, new_value)
        elif choice == "5":
            manager.generate_report()
        elif choice == "6":
            print("Exiting system...")
            break
        else:
            print("Invalid choice. Please enter a number from 1 to 6.")



Inventory Management System
1. Display Inventory
2. Add Product
3. Delete Product
4. Update Product
5. Generate Report
6. Exit
Enter your choice: 2
Enter product name: Laptop
Enter product category: electronics
Enter quantity: 3
Enter price: 56700
Added Laptop with ID 8.

Inventory Management System
1. Display Inventory
2. Add Product
3. Delete Product
4. Update Product
5. Generate Report
6. Exit
Enter your choice: 6
Exiting system...


Prompt 7: Now taking all the input cases provided. Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

In [16]:
import csv
import os
import random

# Constants
CSV_FILE = "inventory.csv"
RESTOCK_THRESHOLD = 5   # If stock < 5, restock automatically
RESTOCK_AMOUNT = 20     # Amount to restock
CATEGORIES = ["Electronics", "Clothing", "Groceries", "Home", "Toys"]

# User authentication data (username and password dictionary)
USERS = {
    "admin": "admin123","manager": "manager456","staff": "staff789"
}

# Random Product Names for Initialization
PRODUCT_NAMES = {
    "Electronics": ["Laptop", "Smartphone", "Tablet", "Headphones", "Monitor"],
    "Clothing": ["T-Shirt", "Jeans", "Jacket", "Sneakers", "Cap"],
    "Groceries": ["Milk", "Bread", "Eggs", "Rice", "Cereal"],
    "Home": ["Sofa", "Table", "Chair", "Lamp", "Curtains"],
    "Toys": ["Action Figure", "Board Game", "Puzzle", "Doll", "RC Car"]
}

# Function to handle user authentication
def authenticate_user():
    print("Please log in to access the inventory system.")
    username = input("Username: ").strip()
    password = input("Password: ").strip()

    if username in USERS and USERS[username] == password:
        print(f"Welcome {username}!")
        return username
    else:
        print("Invalid username or password.")
        return None

# Generate Random Products for Initialization (50 products)
def generate_random_products(num_products=50):
    products = []
    used_ids = set()
    for i in range(num_products):
        category = random.choice(CATEGORIES)
        name = random.choice(PRODUCT_NAMES[category])
        # Ensure unique product IDs during initialization
        while True:
            product_id = random.randint(1001, 9999)  # Avoid starting with 0
            if product_id not in used_ids:
                used_ids.add(product_id)
                break
        stock = random.randint(1, 30)
        price = round(random.uniform(5, 1000), 2)  # Prices between $5 and $1000
        products.append([product_id, name, category, stock, price])
    return products

# Initialize CSV with random data if it doesn't exist
def initialize_inventory():
    if not os.path.exists(CSV_FILE):
        products = generate_random_products()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
            writer.writerows(products)
        print("Inventory initialized with randomly generated data.")

# Load inventory from CSV
def load_inventory():
    try:
        with open(CSV_FILE, "r") as file:
            reader = csv.DictReader(file)
            return list(reader)
    except Exception:
        print("Error reading inventory file. Creating a new one.")
        initialize_inventory()
        return load_inventory()

# Save inventory back to CSV
def save_inventory(inventory):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["ID", "Name", "Category", "Stock", "Price"])
        for item in inventory:
            writer.writerow([item["ID"], item["Name"], item["Category"], item["Stock"], item["Price"]])

# Add new product with validations
def add_product():
    inventory = load_inventory()

    # Input and validate product ID
    product_id_str = input("Enter Product ID (numeric, not starting with 0): ").strip()
    if not product_id_str.isdigit() or product_id_str[0] == "0":
        print("Invalid Product ID. It must be numeric and cannot start with 0.")
        return
    new_id = int(product_id_str)

    # Check if product ID already exists
    for item in inventory:
        if int(item["ID"]) == new_id:
            print("Product ID already exists. Duplicate IDs are not allowed.")
            return

    # Input product name
    name = input("Enter Product Name: ").strip()
    if not name:
        print("Product Name cannot be empty.")
        return

    # Input and validate category
    category = input(f"Enter Category {CATEGORIES}: ").strip()
    if category not in CATEGORIES:
        print("Invalid category. Product not added.")
        return

    # Ensure same product name is not used under a different category
    for item in inventory:
        if item["Name"].lower() == name.lower():
            if item["Category"] != category:
                print(f"Product '{name}' already exists under category '{item['Category']}'.")
                return
            else:
                print(f"Product '{name}' already exists. Duplicate insertion is not allowed.")
                return

    # Input and validate stock
    stock_str = input("Enter Stock Quantity (whole number, non-negative): ").strip()
    if not stock_str.isdigit():
        print("Invalid stock quantity. Please enter a non-negative integer.")
        return
    stock = int(stock_str)

    # Input and validate price
    try:
        price = float(input("Enter Price: ").strip())
    except ValueError:
        print("Invalid price. Please enter a valid number.")
        return

    # Append new product to inventory and save
    inventory.append({
        "ID": new_id,
        "Name": name,
        "Category": category,
        "Stock": stock,
        "Price": price
    })
    save_inventory(inventory)
    print(f"Product '{name}' added successfully!")

# Modify an existing product's details
def modify_product():
    inventory = load_inventory()

    product_id_str = input("Enter the Product ID to modify: ").strip()
    if not product_id_str.isdigit():
        print("Invalid Product ID format.")
        return

    product_id = int(product_id_str)
    product_found = False

    for item in inventory:
        if int(item["ID"]) == product_id:
            product_found = True
            print(f"\nCurrent Details of Product ID {product_id}:")
            print(f"Name: {item['Name']}, Category: {item['Category']}, Stock: {item['Stock']}, Price: ${item['Price']:.2f}")

            # Modify name
            new_name = input(f"Enter new name for the product (or press Enter to keep '{item['Name']}'): ").strip()
            if new_name:
                item['Name'] = new_name

            # Modify category
            new_category = input(f"Enter new category for the product (or press Enter to keep '{item['Category']}'): ").strip()
            if new_category and new_category in CATEGORIES:
                item['Category'] = new_category

            # Modify stock
            new_stock_str = input(f"Enter new stock quantity for the product (or press Enter to keep {item['Stock']}): ").strip()
            if new_stock_str:
                if new_stock_str.isdigit():
                    item['Stock'] = int(new_stock_str)
                else:
                    print("Invalid stock quantity.")

            # Modify price
            new_price_str = input(f"Enter new price for the product (or press Enter to keep ${item['Price']}): ").strip()
            if new_price_str:
                try:
                    item['Price'] = float(new_price_str)
                except ValueError:
                    print("Invalid price entered.")

            save_inventory(inventory)
            print(f"Product ID {product_id} updated successfully!")
            break

    if not product_found:
        print("Product ID not found. No modification made.")

# Delete product with validation for non-existing product
def delete_product():
    inventory = load_inventory()
    product_id = input("Enter Product ID to delete: ").strip()

    # Validate product_id input
    if not product_id.isdigit():
        print("Invalid Product ID format.")
        return

    found = False
    new_inventory = []
    for item in inventory:
        if str(item["ID"]) == product_id:
            found = True
        else:
            new_inventory.append(item)

    if not found:
        print("Product ID not found. No deletion performed.")
    else:
        save_inventory(new_inventory)
        print("Product deleted successfully.")

# Search product by name or category
def search_product():
    inventory = load_inventory()
    search_term = input("Enter product name or category to search: ").lower().strip()
    results = [item for item in inventory if search_term in item["Name"].lower() or search_term in item["Category"].lower()]

    if results:
        print("\nSearch Results:")
        for item in results:
            print(item)
    else:
        print("No matching products found.")

# Restock items automatically if stock is below threshold
def auto_restock():
    inventory = load_inventory()
    restocked_items = []

    for item in inventory:
        if int(item["Stock"]) < RESTOCK_THRESHOLD:
            new_stock = int(item["Stock"]) + RESTOCK_AMOUNT
            item["Stock"] = new_stock
            restocked_items.append(item["Name"])

    save_inventory(inventory)

    if restocked_items:
        print("Restocked:", ", ".join(restocked_items))
    else:
        print("No restocking needed.")

# Generate report with an option for a detailed product-level report
def generate_report():
    inventory = load_inventory()
    report_type = input("Enter report type (summary / detailed): ").strip().lower()

    if report_type == "detailed":
        print("\n=== Detailed Product Report ===")
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            print(f"ID: {item['ID']}, Name: {item['Name']}, Category: {item['Category']}, "
                  f"Stock: {item['Stock']}, Price: ${float(item['Price']):.2f}, Value: ${value:.2f}")
    else:
        category_values = {}
        total_value = 0
        for item in inventory:
            value = int(item["Stock"]) * float(item["Price"])
            total_value += value
            category = item["Category"]
            category_values[category] = category_values.get(category, 0) + value

        print("\n=== Inventory Report ===")
        print(f"Total Inventory Value: ${total_value:.2f}")
        print("Category Breakdown:")
        for category, value in category_values.items():
            print(f"{category}: ${value:.2f}")

# Main menu
def main():
    initialize_inventory()
    # Authenticate user
    username = authenticate_user()
    if username is None:
        print("Access denied. Exiting program.")
        return

    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Delete Product")
        print("3. Modify Product")
        print("4. Search Product")
        print("5. Restock Items")
        print("6. Generate Report")
        print("7. Exit")

        choice = input("Enter your choice: ").strip()

        if choice == "1":
            add_product()
        elif choice == "2":
            delete_product()
        elif choice == "3":
            modify_product()
        elif choice == "4":
            search_product()
        elif choice == "5":
            auto_restock()
        elif choice == "6":
            generate_report()
        elif choice == "7":
            print("Exiting system")
            break
        else:
            print("Invalid choice. Try again.")

# Run the program
if __name__ == "__main__":
    main()

Please log in to access the inventory system.
Username: manager
Password: manager456
Welcome manager!

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 1
Enter Product ID (numeric, not starting with 0): 8
Product ID already exists. Duplicate IDs are not allowed.

Inventory Management System
1. Add Product
2. Delete Product
3. Modify Product
4. Search Product
5. Restock Items
6. Generate Report
7. Exit
Enter your choice: 7
Exiting system


# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

Prompt 1: Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

In [17]:
import random
import csv
import os

# Constants
CSV_FILE = "flights.csv"
SEAT_CAPACITY = 150  # Default seat capacity for each flight

# Generate 50 flights with random data
def generate_flights():
    flights = []
    cities = ["New York", "Los Angeles", "Chicago", "Houston", "Miami", "Seattle", "San Francisco", "Denver", "Boston", "Las Vegas"]
    for i in range(50):
        flight_id = f"F{random.randint(1000, 9999)}"
        departure = random.choice(cities)
        destination = random.choice([c for c in cities if c != departure])
        seats_available = random.randint(50, SEAT_CAPACITY)
        price = round(random.uniform(100, 1000), 2)
        flights.append([flight_id, departure, destination, seats_available, price])
    return flights

# Initialize CSV file with flights if not exist
def initialize_flights():
    if not os.path.exists(CSV_FILE):
        flights = generate_flights()
        with open(CSV_FILE, "w", newline="") as file:
            writer = csv.writer(file)
            writer.writerow(["FlightID", "Departure", "Destination", "SeatsAvailable", "Price"])
            writer.writerows(flights)
        print("Flight data initialized.")

# Load flights from CSV
def load_flights():
    with open(CSV_FILE, "r") as file:
        reader = csv.DictReader(file)
        return list(reader)

# Save flights back to CSV
def save_flights(flights):
    with open(CSV_FILE, "w", newline="") as file:
        writer = csv.writer(file)
        writer.writerow(["FlightID", "Departure", "Destination", "SeatsAvailable", "Price"])
        for flight in flights:
            writer.writerow([flight["FlightID"], flight["Departure"], flight["Destination"], flight["SeatsAvailable"], flight["Price"]])

# Search for available flights
def search_flights():
    departure = input("Enter departure city: ").strip()
    destination = input("Enter destination city: ").strip()
    flights = load_flights()

    results = [f for f in flights if f["Departure"].lower() == departure.lower() and f["Destination"].lower() == destination.lower()]

    if results:
        print("Available Flights:")
        for flight in results:
            print(f"Flight {flight['FlightID']} - Seats: {flight['SeatsAvailable']}, Price: ${flight['Price']}")
    else:
        print("No flights available for the selected route.")

# Book a seat on a flight
def book_flight():
    flight_id = input("Enter Flight ID to book: ").strip()
    flights = load_flights()

    for flight in flights:
        if flight["FlightID"] == flight_id:
            if int(flight["SeatsAvailable"]) > 0:
                flight["SeatsAvailable"] = str(int(flight["SeatsAvailable"]) - 1)
                save_flights(flights)
                print(f"Booking successful! Seat reserved on Flight {flight_id}.")
            else:
                print("Flight is fully booked.")
            return
    print("Flight not found.")

# Cancel a booking
def cancel_booking():
    flight_id = input("Enter Flight ID to cancel booking: ").strip()
    flights = load_flights()

    for flight in flights:
        if flight["FlightID"] == flight_id:
            flight["SeatsAvailable"] = str(int(flight["SeatsAvailable"]) + 1)
            save_flights(flights)
            print(f"Booking canceled. Seat released for Flight {flight_id}.")
            return
    print("Flight not found.")

# Generate a report of flight availability
def generate_report():
    flights = load_flights()
    print("\nFlight Report:")
    for flight in flights:
        print(f"{flight['FlightID']} - {flight['Departure']} to {flight['Destination']} - Available Seats: {flight['SeatsAvailable']}")

# Main Menu
def main():
    initialize_flights()
    while True:
        print("\nFlight Booking System")
        print("1. Search Flights")
        print("2. Book a Flight")
        print("3. Cancel a Booking")
        print("4. Generate Report")
        print("5. Exit")
        choice = input("Select an option: ").strip()

        if choice == "1":
            search_flights()
        elif choice == "2":
            book_flight()
        elif choice == "3":
            cancel_booking()
        elif choice == "4":
            generate_report()
        elif choice == "5":
            print("Exiting system. Goodbye!")
            break
        else:
            print("Invalid option. Please try again.")

if __name__ == "__main__":
    main()
''

Flight data initialized.

Flight Booking System
1. Search Flights
2. Book a Flight
3. Cancel a Booking
4. Generate Report
5. Exit
Select an option: 1
Enter departure city: Mumbai
Enter destination city: Hyderabad
No flights available for the selected route.

Flight Booking System
1. Search Flights
2. Book a Flight
3. Cancel a Booking
4. Generate Report
5. Exit
Select an option: 5
Exiting system. Goodbye!


Prompt 2: Now, by the below provided cases just develop the code accordingly. Like, duplicate flight entries by checking for existing flight IDs before insertion. The system ensures that the same flight route (origin and destination) cannot be added under different airline categories. Flight IDs must contain only numbers and should not start with zero. Additionally, negative seat availability values are restricted to maintain data accuracy, and seat reservations require whole numbers, disallowing float values. The report generation feature now includes an additional level, allowing users to generate reports for individual flights if needed. Users are also informed if they attempt to update or delete a non-existent flight. Moreover, flight names cannot be empty, ensuring meaningful entries, and duplicate flight IDs are strictly avoided to prevent inconsistencies

In [5]:
# Your code here
import csv
import os
import random

# Filepath for flights CSV file
FLIGHTS_FILE = "flights.csv"

# Initialize the flight system
def initialize_flights():
    if not os.path.exists(FLIGHTS_FILE):
        with open(FLIGHTS_FILE, mode='w', newline='') as file:
            writer = csv.writer(file)
            writer.writerow(["Flight ID", "Airline", "Origin", "Destination", "Seats Available", "Total Seats", "Price"])
        populate_initial_flights()

# Load flights from CSV
def load_flights():
    flights = []
    try:
        with open(FLIGHTS_FILE, mode='r') as file:
            reader = csv.DictReader(file)
            for row in reader:
                flights.append(row)
    except Exception:
        print("Error reading flight file. Initializing a new flight system.")
        initialize_flights()
    return flights

# Add a new flight
def add_flight(flight_id, airline, origin, destination, seats_available, total_seats, price):
    flights = load_flights()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            print("Error: Flight ID already exists.")
            return
    with open(FLIGHTS_FILE, mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerow([flight_id, airline, origin, destination, seats_available, total_seats, price])
    print("Flight added successfully.")

# Find flights by origin and destination
def find_flight(origin, destination):
    flights = load_flights()
    results = [flight for flight in flights if flight["Origin"].lower() == origin.lower() and flight["Destination"].lower() == destination.lower()]
    return results

# Reserve a seat
def reserve_seat(flight_id):
    flights = load_flights()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Seats Available"]) > 0:
                flight["Seats Available"] = str(int(flight["Seats Available"]) - 1)
                save_flights(flights)
                print("Seat reserved successfully.")
                return
            else:
                print("Error: No available seats.")
                return
    print("Error: Flight ID not found.")

# Release a seat (Cancellation)
def release_seat(flight_id):
    flights = load_flights()
    for flight in flights:
        if flight["Flight ID"] == flight_id:
            if int(flight["Seats Available"]) < int(flight["Total Seats"]):
                flight["Seats Available"] = str(int(flight["Seats Available"]) + 1)
                save_flights(flights)
                print("Seat cancellation successful.")
                return
            else:
                print("Error: Cannot release more seats than total available.")
                return
    print("Error: Flight ID not found.")

# Generate performance reports
def generate_report():
    flights = load_flights()
    total_revenue = sum(int(flight["Total Seats"]) * float(flight["Price"]) for flight in flights)
    print("Total Revenue from Flights: $", round(total_revenue, 2))
    print("Flight Summary:")
    for flight in flights:
        print(f"{flight['Airline']} | {flight['Origin']} -> {flight['Destination']} | Seats Left: {flight['Seats Available']}")

# Save updated flight records
def save_flights(flights):
    with open(FLIGHTS_FILE, mode='w', newline='') as file:
        writer = csv.DictWriter(file, fieldnames=["Flight ID", "Airline", "Origin", "Destination", "Seats Available", "Total Seats", "Price"])
        writer.writeheader()
        writer.writerows(flights)

# Populate initial flights with 50 different records
def populate_initial_flights():
    airlines = ["Delta", "United", "American Airlines", "Southwest", "JetBlue"]
    cities = ["New York", "Los Angeles", "Chicago", "Houston", "Miami", "San Francisco"]
    sample_flights = [
        (f"F{100+i}", random.choice(airlines), random.choice(cities), random.choice([c for c in cities if c != cities[i % len(cities)]]), random.randint(10, 200), 200, round(random.uniform(100, 1000), 2))
        for i in range(50)
    ]
    with open(FLIGHTS_FILE, mode='a', newline='') as file:
        writer = csv.writer(file)
        writer.writerows(sample_flights)

# Ensure flight system maintains at least 50 flights
def maintain_flights():
    flights = load_flights()
    if len(flights) < 50:
        populate_initial_flights()

# Initialize flight system and maintain 50 flights
initialize_flights()
maintain_flights()

# Test functions
add_flight("F150", "Alaska Airlines", "Boston", "Seattle", 50, 150, 350.00)
reserve_seat("F100")
release_seat("F100")
print("Flight Search Results:", find_flight("New York", "Los Angeles"))
generate_report()



Flight added successfully.
Seat reserved successfully.
Seat cancellation successful.
Flight Search Results: [{'Flight ID': 'F127', 'Airline': 'JetBlue', 'Origin': 'New York', 'Destination': 'Los Angeles', 'Seats Available': '117', 'Total Seats': '200', 'Price': '972.91'}]
Total Revenue from Flights: $ 5830586.0
Flight Summary:
JetBlue | Miami -> Chicago | Seats Left: 48
United | New York -> San Francisco | Seats Left: 11
Southwest | New York -> New York | Seats Left: 105
United | Houston -> San Francisco | Seats Left: 115
Southwest | New York -> Houston | Seats Left: 115
United | Miami -> Chicago | Seats Left: 67
United | Chicago -> Miami | Seats Left: 80
Southwest | New York -> New York | Seats Left: 156
United | San Francisco -> New York | Seats Left: 110
American Airlines | Los Angeles -> New York | Seats Left: 188
Southwest | Los Angeles -> San Francisco | Seats Left: 58
Delta | Houston -> Miami | Seats Left: 148
JetBlue | Chicago -> Houston | Seats Left: 191
American Airlines | Ne

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

# Write your response below
I felt the 6th and 7th one could be the easiest.
But While learning or handing the edge cases the real struggle starts.
 Yeah!!I could Say its really Brainstorming and felt like learnt something new!!!